MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Services.Tests/Services/ListCacheTests.cs

```diff
@@ -0,0 +1,190 @@
+            MockCachedItemsProviderItemsFor<ICollection<ListItem>>();
+        }
+
+        private void MockCachedItemsProviderItemsFor<T>()
+            => _cachedItemsProvider
+                .ExecuteOnItems(Arg.Any<Func<ConcurrentDictionary<Guid, ListItem>, T>>())
+                .Returns(callInfo => callInfo.Arg<Func<ConcurrentDictionary<Guid, ListItem>, T>>().Invoke(_items));
+
+        [Test]
+        public async Task AddItemAsync_item_AddItemIntoRepository()
+        {
+            var stringId = "22AC59B7-9517-4EDD-9DDD-EB418A7C1678";
+            var newItem = ListItemBuilder.CreateItem(stringId, "newItem");
+            var id = Guid.Parse(stringId);
+            _listRepository.AddItemAsync(newItem).Returns(newItem);
+
+            var addedItem = await _listCache.AddItemAsync(newItem);
+
+            Assert.Multiple(async () => {
+                await _listRepository.Received(1).AddItemAsync(newItem);
+                Assert.That(addedItem, Is.EqualTo(newItem));
+            });
+        }
+
+        [Test]
+        public async Task DeleteItemAsync_itemId_DeleteItemFromRepository()
+        {
+            var id = _items.Values.First().Id;
+            _listRepository.DeleteItemAsync(id).Returns(_items[id]);
+
+            var deletedItem = await _listCache.DeleteItemAsync(id);
+
+            Assert.Multiple(async () => {
+                await _listRepository.Received(1).DeleteItemAsync(id);
+                Assert.That(deletedItem, Is.Not.Null.And.Property(nameof(ListItem.Id)).EqualTo(id));
+            });
+        }
+
+        [Test]
+        public void DeleteItemAsync_itemId_ThrowsException()
+        {
+            var id = new Guid("22AC59B7-9517-4EDD-9DDD-EB418A7C1678");
+
+            Assert.Multiple(async () => {
+                Assert.ThrowsAsync<ArgumentException>(async () => await _listCache.DeleteItemAsync(id));
+
+                await _listRepository.DidNotReceive().DeleteItemAsync(Arg.Any<Guid>());
+            });
+        }
+
+        [Test]
+        public async Task ReplaceItemAsync_item_ReplaceItemInRepository()
+        {
+            var editedItem = _items.Values.First().With(item => item.Text, "bbbbb");
+
+            await _listCache.ReplaceItemAsync(editedItem);
+
+            Assert.Multiple(async () => {
+                await _listRepository.Received(1).ReplaceItemAsync(editedItem);
+            });
+        }
+
+        [Test]
+        public void ReplaceItemAsync_NonexistentItem_ThrowsException()
+        {
+            var item = ListItemBuilder.CreateItem("22AC59B7-9517-4EDD-9DDD-EB418A7C1678");
+
+            Assert.Multiple(async () => {
+                Assert.ThrowsAsync<ArgumentException>(async () => await _listCache.ReplaceItemAsync(item));
+
+                await _listRepository.DidNotReceive().ReplaceItemAsync(Arg.Any<ListItem>());
+            });
+        }
+
+        [Test]
+        public async Task GetAllItemsAsync_ReturnsAllItems()
```

Comment:
Why do you call these methods twice, now? I mean, since you split `CachedItemProvider` from `ListCache`, there is a no reason for these methods to be called twice - it is work for `CachedItemProviderTests` (unit level) and `ApplicationTests` (integration level) now, isn't?